src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+    VirtioSndRJackRemap,
+
+    /* PCM control request types */
+    VirtioSndRPcmInfo = 0x0100,
+    VirtioSndRPcmSetParams,
+    VirtioSndRPcmPrepare,
+    VirtioSndRPcmRelease,
+    VirtioSndRPcmStart,
+    VirtioSndRPcmStop,
+
+    /* channel map control request types */
+    VirtioSndRChmapInfo = 0x0200,
+
+    /* jack event types */
+    VirtioSndEvtJackConnected = 0x1000,
+    VirtioSndEvtJackDisconnected,
+
+    /* PCM event types */
+    VirtioSndEvtPcmPeriodElapsed = 0x1100,
+    VirtioSndEvtPcmXrun,
+
+    /* common status codes */
+    /// success
+    #[num_enum(default)]
+    VirtioSndSOk = 0x8000,
+    /// a control message is malformed or contains invalid parameters
+    VirtioSndSBadMsg,
+    /// requested operation or parameters are not supported
+    VirtioSndSNotSupp,
+    ///  an I/O error occurred
+    VirtioSndSIoErr,
+}
+
+/// Enum representing the types of item information requests.
+#[derive(Clone, Copy)]
+pub enum ItemInfomationRequestType {
+    /// Represents a jack information request.
+    VirtioSndRJackInfo = 1,
+    /// Represents a PCM information request.
+    VirtioSndRPcmInfo = 0x0100,
+    /// Represents a channel map information request.
+    VirtioSndRChmapInfo = 0x0200,
+}
+
+impl From<ItemInfomationRequestType> for VirtIOSndHdr {
+    fn from(value: ItemInfomationRequestType) -> Self {
+        VirtIOSndHdr {
+            command_code: value as _,
+        }
+    }
+}
+
+impl Into<u32> for ItemInfomationRequestType {
+    fn into(self) -> u32 {
+        self as _
+    }
+}
+
+#[derive(IntoPrimitive)]
+#[repr(u32)]
+enum RequestStatusCode {
+    /* common status codes */
+    VirtioSndSOk = 0x8000,
+    VirtioSndSBadMsg,
+    VirtioSndSNotSupp,
+    VirtioSndSIoErr,
+}
+
+impl From<RequestStatusCode> for VirtIOSndHdr {
+    fn from(value: RequestStatusCode) -> Self {
+        VirtIOSndHdr {
+            command_code: value as _,
+        }
+    }
+}
+
+/// A common header
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes, PartialEq, Eq)]
+struct VirtIOSndHdr {
+    command_code: u32,
+}
+
+impl From<CommandCode> for VirtIOSndHdr {
+    fn from(value: CommandCode) -> Self {
+        VirtIOSndHdr { command_code:  value.into() }
+    }
+}
+
+#[repr(C)]
+#[derive(FromBytes, FromZeroes)]
+/// An event notification
+struct VirtIOSndEvent {
+    hdr: VirtIOSndHdr,
+    data: u32,
+}
+
+/// The notification type.
+#[repr(u32)]
+#[derive(FromPrimitive, Copy, Clone)]
+pub enum NotificationType {
+    /// A hardware buffer period has elapsed, the period size is controlled using the `period_bytes` field.
+    #[num_enum(default)]
+    PcmPeriodElapsed = 0x1100,
+    /// An underflow for the output stream or an overflow for the inputstream has occurred.
+    PcmXrun,
+    /// An external device has been connected to the jack.
+    JackConnected = 0x1000,
+    /// An external device has been disconnected from the jack.
+    JackDisconnected,
+}
+
+/// Notification from sound device.
+pub struct Notification {
```

Comment:
Derive `Clone`, `Debug`, `Eq` and `PartialEq` as users of the library may find them useful.

Reply:
Added.